File: server/php/src/Payment/CreatePaymentController.php
<?php
namespace MyApp\Payment;

use Laminas\Diactoros\Response\JsonResponse;
use MyApp\Payment\CreatePaymentMapper;
use InvalidArgumentException;
use Slim\Psr7\Request;
use Slim\Psr7\Response;

class CreatePaymentController
{
    private $mapper;
    private $createPaymentService;
    private $paymentDetailsService;

    public function __construct(
        CreatePaymentMapper $mapper,
        CreatePaymentService $createPaymentService,
        PaymentDetailsService $paymentDetailsService
    ) {
        $this->mapper = $mapper;
        $this->createPaymentService = $createPaymentService;
        $this->paymentDetailsService = $paymentDetailsService;
    }

    public function initializeRequest(Request $request, Response $response, $args)
    {
        $dto = $this->mapper->toEmptyBasicDto();
        return new JsonResponse($dto, 200, [], true);
    }

    public function getPaymentResponse(Request $request, Response $response, $args)
    {
        $paymentId = $request->getQueryParams()['paymentId'];

        if (empty($paymentId)) {
            throw new InvalidArgumentException("Payment id is missing!");
        }

        $paymentDetailsResponse = $this->paymentDetailsService->getPaymentDetails($paymentId);
        return new JsonResponse($paymentDetailsResponse->toObject(), 200, [], true);
    }

    public function createPaymentRequest(Request $request, Response $response, $args)
    {
        $createPaymentRequest = $this->mapper->toCreatePaymentRequest($this->mapper->toCreatePaymentBasicDto($request));
        $createPaymentResponse = $this->createPaymentService->createPayment($createPaymentRequest);

        return new JsonResponse($createPaymentResponse->toObject(), 200, [], true);  
    }

}

File: server/python/payment/create_payment_controller.py
from flask import request # type: ignore
from payment import create_payment_mapper, create_payment_service, payment_details_service

def initialize_payment():
    return create_payment_mapper.to_empty_dto()

def create_payment(payment_request_dto):
    create_payment_request = create_payment_mapper.to_payment_request(payment_request_dto)
    create_payment_response = create_payment_service.create_payment(create_payment_request)
    return create_payment_response.to_dictionary()

def outcome():
    payment_id = request.args.get('paymentId')
    payment_details_response = payment_details_service.get_payment_details(payment_id)
    return payment_details_response.to_dictionary()